docs/RFCS/split_sql_syntax.md

```diff
@@ -0,0 +1,138 @@
+- Feature Name: SQL Syntax for splitting tables
+- Status: draft/in-progress/completed/rejected/obsolete
+- Start Date: 2017-03-14
+- Authors: Radu Berinde
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #13665
+
+# Summary
+
+This RFC proposes new syntax for introducing table splits via SQL.
+
+# Motivation
+
+The main motivation is to allow setting up tests, benchmarks, and reproducible
+testbeds, especially for DistSQL. One set of some sample tests that we want to
+be able to set up directly from logic tests are in
+[TestDistSQLPlanner](https://github.com/cockroachdb/cockroach/blob/cc5ba57/pkg/sql/distsql_physical_planner_test.go#L75).
+These tests need to split tables in a specific way, and they need to reprogram
+the replicas and range leaders of each split.
+
+The secondary motivation is for backup/restore, which needs to introduce a
+specific set of key splits. Currently this is done by custom code which works as
+follows: the keys are sorted; a split is introduced on the middle key, then
+the splits to the left and respectively to the right are processed recursively
+(in parallel). We want the new syntax to implement this algorithm so that
+backup/restore can use it. Note that in this case, we already have the split keys
+(as opposed to primary key values). In addition to backup/restore, there are
+more usecases for pre-spliting ranges (for benchmarks, importing, etc).
+
+We currently have the `ALTER TABLE/INDEX SPLIT AT` statement which allows
+introducing a table or index split at specific column values. The main drawback
+is that it cannot be used *programmatically*: each split must be in its own
+statement; we cannot generate (e.g. via a `SELECT`) a set of values where to
+split. We want to write tests like the one linked above without hardcoding each
+split in the test file; and we want to be able to easily change the test table
+sizes. In addition, the `SPLIT AT` statments don't support control of
+replication.
+
+
+# Detailed design
+
+Three new pieces of SQL syntax are proposed.
+
+### 1. `key_prefix` function ###
+
+The `key_prefix` function is a function that takes a table name (optionally with
+an index) and a set of column values for a prefix of the primary key (or index
+columns) and returns the corresponding key (as `BYTES`). Example:
+
+```sql
+CREATE TABLE t (
+  k1 INT,
+  k2 INT,
+  v INT,
+  w INT,
+  PRIMARY KEY(k1, k2),
+  INDEX vw (v, w)
+)
+
+SELECT key_prefix('t', 1, 2)     -- returns /t/primary/1/2
+SELECT key_prefix('db.t', 3)     -- returns /t/primary/3
+SELECT key_prefix('t@vw', 4, 5)  -- returns /t/vw/4/5
+SELECT key_prefix('t@vw', 1)     -- returns /t/vw/1
+```
+
+#### `pretty_key` function ####
+
+For debugging/testing purposes, we will also need a `pretty_key` function that
+takes a `BYTES` key generated by `key_prefix` and returns a pretty-printed string.
+
+### 2. `KVSPLIT` statement ###
```

Comment:
I would recommend having these two statements as a special form of `ALTER`. I have seen clients classify DDL vs non-DDL statemetns by looking at the first word, and a new word is asking for trouble.

Probably this new `KVSPLIT` will simply end up replacing our current `ALTER ... SPLIT`. I think this would be reasonable. Like Peter suggests, our current SPLIT node can be extended with a SELECT clause, the grammar allows this and the corresponding changes would not be too complicated.